src/test/AutoRoller.t.sol

```diff
@@ -936,12 +978,195 @@ contract AutoRollerTest is DSTestPlus {
+
+        // Check that the Target dust leftover is small
+        assertLt(target.balanceOf(address(autoRoller)), 1e9);
+    }
+
+    function testMaxRedeemDuringCooldown(uint256 amount) public {
+        // 1. Make a deposit during cooldown.
+        uint256 sharesOut = autoRoller.deposit(0.1e18, address(this));
+
+        // 2. Assert max withdraw is same sharesOut.
+        assertEq(autoRoller.maxRedeem(address(this)), sharesOut);
+
+        // 4. Roll into the first Series.
+        autoRoller.roll();
+
+        // 3. Make another deposit during cooldown.
+        sharesOut += autoRoller.deposit(0.1e18, address(this));
+
+        vm.warp(autoRoller.maturity());
+        
+        // 4. Increase scale and settle the first Series.
+        autoRoller.settle();
+
+        // 5. Assert max withdraw is same total shares.
+        assertEq(autoRoller.maxRedeem(address(this)), sharesOut);
+    }
+
+    function testMaxRedeemWithPTsExcess(uint256 amount) public {
+        // 1. Make a deposit durring cooldown.
+        uint256 sharesOut = autoRoller.deposit(1e18, address(this));
+
+        // 2. Roll into the first Series.
+        autoRoller.roll();
+
+        // 3. Swap PTs in to generate excess of PTs.
+        target.mint(address(this), 1e18);
+        target.approve(address(divider), 1e18);
+        divider.issue(address(mockAdapter), autoRoller.maturity(), 1e18);
+        ERC20 pt = ERC20(divider.pt(address(mockAdapter), autoRoller.maturity()));
+        Space space = Space(spaceFactory.pools(address(mockAdapter), autoRoller.maturity()));
+        pt.approve(address(balancerVault), 1e18);
+
+        // 4. Swap PTs in to generate excess of PTs and that the number of PTs we can sell
+        // is less than the diff between YTs and PTs.
+        _swap(
+            BalancerVault.SingleSwap({
+                poolId: space.getPoolId(),
+                kind: BalancerVault.SwapKind.GIVEN_IN,
+                assetIn: address(pt),
+                assetOut: address(autoRoller.asset()),
+                amount: 0.001e18, // TODO: how do I calculate a number such that it would make the maxPTSale < diff? This number works though
+                userData: hex""
+            })
+        );
+
+        // 5. Assert max redeem is same as sharesOut.
+        assertEq(autoRoller.maxRedeem(address(this)), sharesOut);
+
+        // 6. Swap PTs in to generate excess of PTs and that the number of PTs we can sell
+        // is more than the diff between YTs and PTs.
+        _swap(
+            BalancerVault.SingleSwap({
+                poolId: space.getPoolId(),
+                kind: BalancerVault.SwapKind.GIVEN_IN,
+                assetIn: address(pt),
+                assetOut: address(autoRoller.asset()),
+                amount: 0.01e18, // TODO: how do I calculate a number such that it would make the maxPTSale >= diff? This number works though
```

Comment:
So, the only case where this isn't true IIRC is if order we'd have to swap here such that the pool becomes very close to the `maxRate` value (time shifted) so that the user can't safely make the PT in swap needed to withdraw their position without exceeding the max rate

Note that you can check the rate "(pt reserves + total supply / (target reserve * scale))"